### PR TITLE
Further cleanup of the side-effect machinery

### PR DESCRIPTION
**Pull request by ppedrot**
This PR is a follow-up of #10151 and is twofold:
- First part sorts out the higher mess in `declare.ml` and removes a hack based on mixing side-effect declaration with normal constant declaration. Now we clearly separate the logical declaration, that happens in one block (and should be moved somewhere else) and the libobject declaration that happens after. There is still a (now documented) hack based on the discharge function to store the value of a section definition in the object to retrieve it after section closure, but unfortunately this is the way of the libobject API.
- Second part forces the construction of a side-effect to occur at the same time as its logical declaration by grouping the calls in the API. We leverage this invariant to remove pseudo-accesses to the opaque body while it is actually stactically known that these bodies are *not* opaque *yet*.

One invariant that is not statically enforced but should be somehow is that section definitions are never indirect opaques. Not sure how to do that by typing though.

Further cleanups lie ahead, but I want to do this by small steps, otherwise reviewing and discussing the PRs ends up eating a lot of time, and this is a pretty sensitive area of the kernel.

### CODE REVIEW

File: interp/declare.ml [hunk 1]
```diff
@@ -153,7 +149,9 @@ let definition_entry ?fix_exn ?(opaque=false) ?(inline=false) ?types
     const_entry_feedback = None;
     const_entry_inline_code = inline}
 
-let declare_constant ?(internal = UserIndividualRequest) ?(local = false) id ?(export_seff=false) (cd, kind) =
+(* TODO: move me somewhere in the STM. *)
```
Comment (SkySkimmer) [inline interp/declare.ml]: A voice told me this is a @maximedenes plan, can you explain more? I have no clue what that would look like.

Comment (ppedrot) [reply interp/declare.ml]: @maximedenes that's your plan, I summon you to defend it. Otherwise I'll just remove the TODO.

Comment (SkySkimmer) [reply interp/declare.ml]: TBH I'm not sure of the value of leaving the TODO even with the plan defended. A TODO with local implications on the code OK sense, but if there's a need for a plan maybe it should live as an issue or cep.

**Comment by ppedrot** [comment]
@SkySkimmer Do you want to assign? I am starting to have a chain of depending PRs, and I don't want to make this one more complex that it is already.

File: kernel/term_typing.mli [hunk 2]
```diff
@@ -43,6 +43,4 @@ val infer_declaration : trust:'a trust -> env ->
   'a constant_entry -> Opaqueproof.proofterm Cooking.result
 
 val build_constant_declaration :
-  force:('a -> constr * 'b) ->
-  iter:((constr -> unit) -> 'a -> 'a) ->
-  env -> 'a Cooking.result -> 'a constant_body
+  env -> Opaqueproof.proofterm Cooking.result -> Opaqueproof.proofterm constant_body
```
Comment (SkySkimmer) [inline kernel/term_typing.mli]: This seems used only internally, why export it?

Comment (ppedrot) [reply kernel/term_typing.mli]: I am not the one who exported it, and I'd be happy to remove it. I suspect there are still people using the trace mechanism of the interpreter to debug, which requires mli declarations.

File: kernel/safe_typing.ml [hunk 3]
```diff
@@ -771,35 +742,41 @@ let export_side_effects mb env c =
...
-               let ce = constant_entry_of_side_effect ocb u in
+            let env, cb =
+              let kn = eff.seff_constant in
+               let ce = constant_entry_of_side_effect eff in
                let cb = Term_typing.translate_constant Term_typing.Pure env kn ce in
-               let eff = { eff with
-                seff_body = cb;
-                seff_env = `Nothing;
-               } in
+               let cb = map_constant (fun c -> Future.force c) cb in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: Why eta expand?

File: kernel/safe_typing.ml [hunk 4]
```diff
@@ -811,9 +788,9 @@ let add_constant ?role ~in_section l decl senv =
         Term_typing.translate_constant (Term_typing.SideEffects handle) senv.env kn ce
       | ConstantEntry (PureEntry, ce) ->
         Term_typing.translate_constant Term_typing.Pure senv.env kn ce
-      | GlobalRecipe r ->
-        let cb = Term_typing.translate_recipe ~hcons:(not in_section) senv.env kn r in
-        if in_section then cb else Declareops.hcons_const_body cb in
+    in
+  let senv =
+    let cb = map_constant (fun c -> Opaqueproof.create c) cb in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: Why eta expand?

File: kernel/safe_typing.ml [hunk 5]
```diff
@@ -771,35 +742,41 @@ let export_side_effects mb env c =
...
   (ce, exported), senv
 
-let add_constant ?role ~in_section l decl senv =
+let add_recipe ~in_section l r senv =
   let kn = Constant.make2 senv.modpath l in
   let senv =
+    let cb =
+      let cb = Term_typing.translate_recipe ~hcons:(not in_section) senv.env kn r in
+      if in_section then cb else Declareops.hcons_const_body cb in
+    add_constant_aux ~in_section senv (kn, cb) in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: Not sure there's a point in having this nested letin structure here.

File: kernel/safe_typing.ml [hunk 6]
```diff
@@ -824,7 +801,16 @@ let add_constant ?role ~in_section l decl senv =
   in
   let eff = match role with
   | None -> empty_private_constants
-  | Some role -> private_constant senv role kn
+  | Some role ->
+    let cb = map_constant (fun c -> Future.force c) cb in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: Why eta expand?

**Review by SkySkimmer — APPROVED** [review]
We need to decide if we keep that TODO comment, otherwise good to go.

File: kernel/term_typing.mli (hunk 2 above)
Comment (SkySkimmer) [reply kernel/term_typing.mli]: Maybe they should compile without mlis, everything transparent

Comment (ppedrot) [reply kernel/term_typing.mli]: Yep, maybe dune can help there.

Comment (ppedrot) [reply kernel/term_typing.mli]: Anyways, I'm not touching this in this PR.

File: kernel/safe_typing.ml [hunk 7]
```diff
@@ -797,7 +797,7 @@ let export_private_constants ~in_section ce senv =
   let senv = List.fold_left (add_constant_aux ~in_section) senv bodies in
   (ce, exported), senv
 
-let add_constant ~in_section l decl senv =
+let add_constant ?role ~in_section l decl senv =
```
[the remark this answers is not in the document]
Comment (ppedrot) [reply kernel/safe_typing.ml]: Actually I was even thinking about a dedicated function to add side-effect definitions rather that via a flag, but due to the global API, that duplicates everything. I can backtrack on it though.

File: kernel/safe_typing.ml (hunk 3 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: A reflex of mine in OCaml to be sure to control the arity of the function. Doesn't hurt either, and is future-proof if we start messing with the internal type of opaque definitions.

File: kernel/safe_typing.ml (hunk 5 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: Nope, but that's stratospheric nitpicking, isn't it?

File: kernel/safe_typing.ml (hunk 3 above)
Comment (SkySkimmer) [reply kernel/safe_typing.ml]: >A reflex of mine in OCaml to be sure to control the arity of the function

How does eta expanding do that?

File: kernel/safe_typing.ml (hunk 5 above)
Comment (SkySkimmer) [reply kernel/safe_typing.ml]: It's also trivial to make the code nicer

File: kernel/safe_typing.ml (hunk 3 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: The discipline is to always write fully applied functions, but that's not enforced by the compiler, it's merely a reminder to the reader. I personally cannot stand the point-free style that is commonplace in Haskell for instance, call me a C programmer if you want.

Comment (SkySkimmer) [reply kernel/safe_typing.ml]: It's not partially applied here though

File: kernel/safe_typing.ml (hunk 5 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: Non-trivial because I need to rebase several branches atop of that, but let's do it.

**Comment by ppedrot** [comment]
I've removed the TODO and addressed a few nitpicks, should be ready now.

File: kernel/safe_typing.ml [hunk 8]
```diff
@@ -771,35 +742,39 @@ let export_side_effects mb env c =
...
+  let bodies = List.map map exported in
   let exported = List.map (fun (kn, _, r) -> (kn, r)) exported in
   let senv = List.fold_left (add_constant_aux ~in_section) senv bodies in
   (ce, exported), senv
 
-let add_constant ~in_section l decl senv =
+let add_recipe ~in_section l r senv =
+  let kn = Constant.make2 senv.modpath l in
+  let cb = Term_typing.translate_recipe ~hcons:(not in_section) senv.env kn r in
+  let cb = if in_section then cb else Declareops.hcons_const_body cb in
```
Comment (SkySkimmer) [inline kernel/safe_typing.ml]: I can see it was already that way before the PR, but why do we tell translate_recipe to hcons and then re-hcons?

**Comment by SkySkimmer** [comment]
I'll wait a couple days to give kernel people time to comment (@maximedenes @mattam82 I guess).
I think @gares also has side effect knowledge.

Then I'll merge.

**Review by gares — APPROVED** [review]

**Comment by gares** [comment]
I think you need a hack from a kernel maintainer, that boils down to @maximedenes since @ppedrot is author

File: kernel/safe_typing.ml (hunk 8 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: I have a PR that makes the kernel hashconsing sane, by centralizing in Safe_typing, but I don't want to include it in this one, that would be too much stuff at once.

File: interp/declare.ml (hunk 1 above)
Comment (maximedenes) [reply interp/declare.ml]: I'm afraid I can do much about hearing voices :)

Comment (maximedenes) [reply interp/declare.ml]: It was not a plan, just a potential idea. I agree that a TODO doesn't seem like the right tool here.

Comment (ppedrot) [reply interp/declare.ml]: The TODO has been removed anyways.

File: kernel/safe_typing.ml (hunk 3 above)
Comment (maximedenes) [reply kernel/safe_typing.ml]: The analogy with `C` also makes little sense to me. It is quite the contrary in fact, you'd pass a pointer to the function here, not an anonymous function... Please fix.

File: interp/declare.ml (hunk 1 above)
Comment (maximedenes) [reply interp/declare.ml]: Ah right. Please mark discussions as resolved when they are, it saves time for other reviewers.

File: kernel/safe_typing.ml (hunk 3 above)
Comment (ppedrot) [reply kernel/safe_typing.ml]: *Seriously* guys? Sigh.

**Review by maximedenes — APPROVED** [review]
LGTM, up to the stylistic remarks @SkySkimmer already made.

**Comment by ppedrot** [comment]
Stratospheric nitpickings resolved, this is ready.

File: kernel/safe_typing.ml (hunk 3 above)
Comment (gares) [reply kernel/safe_typing.ml]: If a painter is allowed to have a blue period, maybe a coder should be allowed to have an eta-long period. ;-) Or maybe not...

**Pull request merged**